assets/src/edit-story/components/library/panes/shared/libraryMoveable.js

```diff
@@ -0,0 +1,260 @@
+      return false;
+    }
+    frame.translate = beforeTranslate;
+    if (cloneRef.current) {
+      if (cloneRef.current.style.opacity !== 1 && !activeDropTargetId) {
+        // We're not doing it in `onDragStart` since otherwise on clicking it would appear, too.
+        cloneRef.current.style.opacity = 1;
+      } else if (activeDropTargetId) {
+        // If there's an active drop target, let's hide the clone.
+        cloneRef.current.style.opacity = 0;
+      }
+      cloneRef.current.style.transform = `translate(${beforeTranslate[0]}px, ${beforeTranslate[1]}px)`;
+      // We also have to move the original target ref for snapping to work.
+      targetBoxRef.current.style.transform = `translate(${beforeTranslate[0]}px, ${beforeTranslate[1]}px)`;
+    }
+    handleDrag(inputEvent);
+    return undefined;
+  };
+
+  const getTargetOffset = useCallback(() => {
+    const overlay = overlayRef.current;
+    let offsetX = 0,
+      offsetY = 0;
+    for (
+      let offsetNode = overlay;
+      offsetNode;
+      offsetNode = offsetNode.offsetParent
+    ) {
+      offsetX += offsetNode.offsetLeft;
+      offsetY += offsetNode.offsetTop;
+    }
+    return {
+      offsetX,
+      offsetY,
+    };
+  }, [overlayRef]);
+
+  const onDragStart = ({ set, inputEvent }) => {
+    setDidManuallyReset(false);
+    // Note: we can't set isDragging true here since a "click" is also considered dragStart.
+    set(frame.translate);
+    setIsDragging(true);
+    startEventTracking(inputEvent);
+
+    // Position the clone that's being dragged.
+    const { offsetX, offsetY } = getTargetOffset();
+    const mediaBox = targetBoxRef.current.getBoundingClientRect();
+    const x1 = mediaBox.left - offsetX;
+    const y1 = mediaBox.top - offsetY;
+    cloneRef.current.style.left = `${x1}px`;
+    cloneRef.current.style.top = `${y1}px`;
+  };
+
+  const onDragEnd = ({ inputEvent }) => {
+    if (didManuallyReset) {
+      return false;
+    }
+    if (isMouseUpAClick(inputEvent, eventTracker.current)) {
+      resetMoveable();
+      onClick();
+      return false;
+    }
+    // We only skip Moveable onDragEnd handling if there's an active drop target ID.
+    if (activeDropTargetId) {
+      handleDragEnd();
+      // Only continue if the clone is at least partially on the page.
+    } else if (!isTargetOutOfContainer(cloneRef.current, pageContainer)) {
+      const {
+        x,
+        y,
+        width: w,
+        height: h,
+      } = cloneRef.current.getBoundingClientRect();
+      const { x: pageX, y: pageY } = pageContainer.getBoundingClientRect();
+
+      insertElement(type, {
+        ...elementProps,
+        x: editorToDataX(x - pageX, pageSize.width),
+        y: editorToDataY(y - pageY, pageSize.height),
+        width: editorToDataX(w, pageSize.width),
+        height: editorToDataY(h, pageSize.height),
+      });
+    }
+    resetMoveable();
+    return undefined;
+  };
+
+  const { offsetX: snappingOffsetX } = getTargetOffset();
```

Comment:
I think this doesn't work correctly for local library images (but works for third-party content):

![local-library-snap-bug](https://user-images.githubusercontent.com/637548/102074258-7d53c300-3dd2-11eb-9c63-51c2ef079547.gif)

Reply:
Hmm, I'm actually not seeing this issue -- is it possible that you scrolled while dragging for this to happen? In this case, it would be related to #2785.

Reply:
Resolving since snapping will be addressed separately for all elements together.